### PR TITLE
Support long lived graceful restart

### PR DESCRIPTION
**Pull request by ishidawataru**

### CODE REVIEW

File: packet/bgp/bgp.go [hunk 1]
```diff
@@ -556,6 +557,97 @@ func NewCapRouteRefreshCisco() *CapRouteRefreshCisco {
...
+		RestartTime: restartTime,
+	}
+}
+
+type CapLongLivedGracefulRestart struct {
+	DefaultParameterCapability
+	Tuples []*CapLongLivedGracefulRestartTuple
+}
+
+func (c *CapLongLivedGracefulRestart) DecodeFromBytes(data []byte) error {
```
Comment (fujita) [inline packet/bgp/bgp.go]: https://github.com/osrg/gobgp/commit/e847113b526d9389bd939669819c8d56854a93fb

Same issue?

Comment (ishidawataru) [inline packet/bgp/bgp.go]: fixed https://github.com/osrg/gobgp/pull/1125/commits/0bfbe327d9dba53871142c5c186f5de9f174fcdf

File: server/server.go [hunk 2]
```diff
@@ -1765,7 +1765,7 @@ func (s *BgpServer) ShutdownNeighbor(addr string) (err error) {
 	s.mgmtCh <- func() {
 		defer close(ch)
 
-		err = s.resetNeighbor("Neighbor shutdown", addr, bgp.BGP_ERROR_SUB_ADMINISTRATIVE_SHUTDOWN)
+		err = s.resetNeighbor("Neighbor shutdown", addr, bgp.BGP_ERROR_SUB_HARD_RESET)
```
Comment (fujita) [inline server/server.go]: HARD_RESET can be used even when a peer doesn't support GR notification?

Comment (ishidawataru) [inline server/server.go]: nope. fixed.

File: server/fsm.go [hunk 3]
```diff
@@ -544,7 +545,11 @@ func capabilitiesFromConfig(pConf *config.Neighbor) []bgp.ParameterCapabilityInt
 			}
 		}
 		time := c.RestartTime
-		caps = append(caps, bgp.NewCapGracefulRestart(restarting, time, tuples))
+		notification := c.NotificationEnabled
+		caps = append(caps, bgp.NewCapGracefulRestart(restarting, notification, time, tuples))
+		if len(ltuples) > 0 {
```
Comment (fujita) [inline server/fsm.go]: Where's 'ltuples'?

File: server/fsm.go [hunk 4]
```diff
@@ -693,8 +698,11 @@ func (h *FSMHandler) recvMessageWithError() (*FsmMsg, error) {
 					"Subcode": body.ErrorSubcode,
 					"Data":    body.Data,
 				}).Warn("received notification")
-
-				sendToErrorCh(FsmStateReason(fmt.Sprintf("%s %s", FSM_NOTIFICATION_RECV, bgp.NewNotificationErrorCode(body.ErrorCode, body.ErrorSubcode).String())))
+				if body.ErrorCode == bgp.BGP_ERROR_CEASE && body.ErrorSubcode == bgp.BGP_ERROR_SUB_HARD_RESET {
+					sendToErrorCh(FSM_HARD_RESET)
```
Comment (fujita) [inline server/fsm.go]: Only when GR enabled?

Comment (ishidawataru) [inline server/fsm.go]: fixed https://github.com/osrg/gobgp/pull/1125/commits/daeb2a01edf5f59abc7b991e91c7a2a3d3b42ba1

File: server/fsm.go (hunk 3 above)
Comment (ishidawataru) [inline server/fsm.go]: fixed in updated commits.

**Comment by fujita** [comment]
Squashed and pushed.

**Pull request closed without merging**